tiledb/sm/subarray/subarray.cc

```diff
@@ -154,13 +155,129 @@ Subarray& Subarray::operator=(Subarray&& subarray) noexcept {
 /*               API              */
 /* ****************************** */
 
+void Subarray::add_label_range(
+    const DimensionLabelReference& dim_label_ref,
+    Range&& range,
+    const bool read_range_oob_error) {
+  const auto dim_idx = dim_label_ref.dimension_id();
+  if (label_range_subset_[dim_idx].has_value()) {
+    // A label range has already been set on this dimension. Do the following:
+    //  * Check this label is the same label that rangers were already set.
+    if (dim_label_ref.name() != label_range_subset_[dim_idx].name()) {
+      throw StatusException(Status_SubarrayError(
+          "Cannot add label range; Dimension is already to set to use "
+          "dimension label '" +
+          label_range_subset_[dim_idx].name()));
+    }
+  } else {
+    // A label range has not yet been set on this dimension. Do the following:
+    //  * Verify no ranges explicitily set on the dimension.
+    //  * Construct LabelRangeSubset for this dimension label.
+    //  * Clear implicitly set range from the dimension ranges.
+    //  * Update is_default (tracks if the range on the dimension is the default
+    //    value of the entire domain).
+    if (range_subset_[dim_label_ref.dimension_id()]
+            .is_explicitly_initialized()) {
+      throw StatusException(Status_SubarrayError(
+          "Cannot add label range; Dimension '" + std::to_string(dim_idx) +
+          "' already has ranges set to it."));
+    }
+    label_range_subset_[dim_idx] =
+        LabelRangeSubset(dim_label_ref, coalesce_ranges_);
+    range_subset_[dim_label_ref.dimension_id()].clear();
+    is_default_[dim_idx] = false;  // Only need to clear default once.
+  }
+
+  // Must reset the result size and tile overlap
+  est_result_size_computed_ = false;
+  tile_overlap_.clear();
+
+  // Restrict the range to the dimension domain and add.
+  auto&& [error_status, oob_warning] =
+      label_range_subset_[dim_idx].range_subset().add_range(
+          range, read_range_oob_error);
+  if (!error_status.ok()) {
+    throw StatusException(Status_SubarrayError(
+        "Cannot add label range for dimension label '" + dim_label_ref.name() +
+        "'; " + error_status.message()));
+  }
+  if (oob_warning.has_value()) {
+    LOG_WARN(
+        oob_warning.value() + " for dimension label '" + dim_label_ref.name() +
+        "'");
+  }
+}
+
+void Subarray::add_label_range(
+    const std::string& label_name,
+    const void* start,
+    const void* end,
+    const void* stride) {
+  // Check input range data is valid data.
+  if (start == nullptr || end == nullptr) {
```

Comment:
nit for future: we should move all validation calls like this to single-line `ensure_...` functions which take the pointer and an optional error message.